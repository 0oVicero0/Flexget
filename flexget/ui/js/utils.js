'use strict';

String.prototype.startsWith = function(prefix) {
  return this.indexOf(prefix) === 0;
};

function registerFlexModule(module) {
  angular.module('flexgetApp').requires.push(module.name);
}

app.menu = [];
app.routes = [];

function register_menu(href, caption, icon, order) {
  href = '/ui/#' + href;
  app.menu.push({href: href, caption: caption, icon: icon, order: order})
}

function register_route(name, url, controller, template) {
  app.routes.push({
    name: name,
    url: url,
    controller: controller,
    template: template
  });
}