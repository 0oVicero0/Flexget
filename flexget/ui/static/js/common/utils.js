'use strict';

String.prototype.startsWith = function(prefix) {
    return this.indexOf(prefix) === 0;
};

function registerFlexModule(module) {
  angular.module('flexgetApp').requires.push(module.name);
}