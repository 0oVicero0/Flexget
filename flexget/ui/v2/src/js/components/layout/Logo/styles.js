import styled from 'emotion/react';
import theme from 'theme';
import headerImage from 'images/header.png';

export const NavLogo = styled.div`
  background: ${theme.palette.accent[900]} url(${headerImage}) no-repeat center;
  background-size: 17.5rem;
  height: 100%;
  transition: ${theme.transitions.create('width')};
  ${theme.breakpoints.up('sm')} {
    width: ${({ open }) => (open ? '19rem' : '5rem')};
    background-size: ${({ open }) => (open ? '17.5rem' : '24rem')};
  }
`;
