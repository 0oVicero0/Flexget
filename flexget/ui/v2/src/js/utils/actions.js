export const createAction = namespace => (type, payload, {
  ignore = false,
  message,
} = {}) => ({
  type,
  payload,
  error: (payload instanceof Error),
  meta: {
    ignore,
    message,
    namespace,
  },
});

export const createLoading = namespace => type => ({
  type,
  meta: {
    namespace,
    loading: true,
  },
});
