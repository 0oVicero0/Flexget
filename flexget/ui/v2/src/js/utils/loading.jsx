import Loadable from 'react-loadable';
import Loading from 'components/common/Loader';

export function createAsyncComponent(loader) {
  return Loadable({
    loading: Loading,
    loader,
  });
}

