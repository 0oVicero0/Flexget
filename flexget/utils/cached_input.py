import copy
import logging
from flexget.plugin import register_plugin

log = logging.getLogger('cached')


class cached(object):
    """
    Implements transparent caching decorator @cached for inputs.

    Decorator has two parameters

    :name: in which the configuration is present in feeds configuration.
    :key: in which the configuration has the cached resource identifier (ie. url). If the :key: is not
    given or present in the configuration :name: is expected to be a cache name (ie. url)

    Configuration assumptions may make this unusable in some (future) inputs
    """

    cache = {}

    def __init__(self, name, key=None):
        self.name = name
        self.key = key

    def __call__(self, func):

        def wrapped_func(*args, **kwargs):
            # get feed from method parameters
            feed = args[1]
            
            # detect api version
            api_ver = 1
            if len(args) == 3:
                api_ver = 2
            
            if api_ver == 1:
                # get name for a cache from feeds configuration
                if not self.name in feed.config:
                    raise Exception('@cache config name %s is not configured in feed %s' % (self.name, feed.name))
                config = feed.config[self.name]
            else:
                config = args[2]

            log.log(5, 'config: %s' % config)
            log.log(5, 'self.name: %s' % self.name)
            log.log(5, 'self.key: %s' % self.key)
            
            if api_ver == 1:
                if isinstance(config, dict) and self.key in config:
                    name = feed.config[self.name][self.key]
                else:
                    name = feed.config[self.name]
            else:
                if isinstance(config, dict) and self.key in config:
                    name = config[self.key]
                else:
                    name = config

            log.debug('cache name: %s (has: %s)' % (name, ', '.join(self.cache.keys())))

            if name in self.cache:
                # return from the cache
                log.log(5, 'cache hit')
                count = 0
                entries = []
                for entry in self.cache[name]:
                    fresh = copy.deepcopy(entry)
                    entries.append(fresh)
                count += 1
                if count > 0:
                    feed.verbose_progress('Restored %s entries from cache' % count, log)
                return entries
            else:
                log.log(5, 'cache miss')
                # call input event
                response = func(*args, **kwargs)
                # store results to cache
                if api_ver == 1:
                    log.debug('v1 storing to cache %s %s entries' % (name, len(feed.entries)))
                    try:
                        self.cache[name] = copy.deepcopy(feed.entries)
                    except TypeError:
                        # might be caused because of backlog restoring some idiotic stuff, so not neccessarily a bug
                        log.critical('Unable to save feed content into cache, if problem persists longer than a day please report this as a bug')
                else:
                    log.debug('storing to cache %s %s entries' % (name, len(feed.entries)))
                    try:
                        self.cache[name] = copy.deepcopy(response)
                    except TypeError:
                        # might be caused because of backlog restoring some idiotic stuff, so not neccessarily a bug
                        log.critical('Unable to save feed content into cache, if problem persists longer than a day please report this as a bug')
                return response
                    

        return wrapped_func


class CacheClearer(object):

    def on_process_start(self, feed):
        """Internal. Clears the input cache on every process start.
        This is neccessary for webui or otherwise it will only use cache.
        """
        log.debug('clearing cache')
        cached.cache = {}

register_plugin(CacheClearer, 'cache_clearer', builtin=True)
