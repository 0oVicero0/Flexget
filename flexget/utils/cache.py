import hashlib
from pathlib import Path

import requests

from flexget.utils.tools import log


def cached_resource(url, base_dir, force=False, max_size=250, directory='cached_resources'):
    """
    Caches a remote resource to local filesystem. Return a tuple of local file name and mime type, use primarily
    for API/WebUI.

    :param url: Resource URL
    :param force: Does not check for existence of cached resource, fetches the remote URL, ignores directory size limit
    :param max_size: Maximum allowed size of directory, in MB.
    :param directory: Name of directory to use. Default is `cached_resources`
    :return: Tuple of file path and mime type
    """
    mime_type = None
    hashed_name = hashlib.md5(url.encode('utf-8')).hexdigest()
    file_path = Path(base_dir) / directory / hashed_name
    directory = file_path.parent

    if not file_path.exists() or force:
        log.debug('caching %s', url)
        response = requests.get(url)
        response.raise_for_status()
        mime_type = response.headers.get('content-type')
        content = response.content
        directory.mkdir(parents=True, exist_ok=True)

        # Checks directory size and trims if necessary.
        size = dir_size(directory) / (1_024 * 1_024.0)
        if not force:
            while size >= max_size:
                log.debug('directory %s size is over the allowed limit of %s, trimming', size, max_size)
                trim_dir(directory)
                size = dir_size(directory) / (1_024 * 1_024.0)
        file_path.write_bytes(content)

    return str(file_path), mime_type


def dir_size(directory: Path):
    """
    Sums the size of all files in a given dir. Not recursive.

    :param directory: Directory to check
    :return: Summed size of all files in Bytes.
    """
    return sum(file.stat().st_size for file in directory.iterdir())


def trim_dir(directory: Path):
    """
    Removed the least accessed file on a given dir

    :param directory: Directory to check
    """

    def access_time(file: Path):
        return file.stat().st_atime

    files = sorted(directory.iterdir(), key=access_time)
    file_name = directory / files[0]
    log.debug('removing least accessed file: %s', file_name)
    file_name.unlink()
