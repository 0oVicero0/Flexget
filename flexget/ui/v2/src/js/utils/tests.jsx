import React from 'react';
import { MuiThemeProvider } from 'material-ui/styles';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

const mockStore = configureMockStore();

export function themed(component) {
  return (
    <MuiThemeProvider>
      { component }
    </MuiThemeProvider>
  );
}

export function router(component) {
  return (
    <MemoryRouter>
      { component }
    </MemoryRouter>
  );
}

export function provider(component, state = {}) {
  return (
    <Provider store={mockStore(state)}>
      { component }
    </Provider>
  );
}
