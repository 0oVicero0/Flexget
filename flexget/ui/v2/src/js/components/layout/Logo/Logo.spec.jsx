import React from 'react';
import renderer from 'react-test-renderer';
import Logo from 'components/layout/Logo';
import { themed, router } from 'utils/tests';

describe('components/layout/Logo', () => {
  it('renders correctly with sideBarOpen', () => {
    const tree = renderer.create(
      router(themed(<Logo sideBarOpen />)),
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('adds logoMini class without sideBarOpen', () => {
    const tree = renderer.create(
      router(themed(<Logo sideBarOpen={false} />)),
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
