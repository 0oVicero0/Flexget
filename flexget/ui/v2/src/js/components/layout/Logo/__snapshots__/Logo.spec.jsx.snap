// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/layout/Logo adds logoMini class without sideBarOpen 1`] = `
<a
  href="/"
  onClick={[Function]}
>
  <div
    className="css-NavLogo-pmb2010 css-1oup17m"
    open={false}
  />
</a>
`;

exports[`components/layout/Logo renders correctly with sideBarOpen 1`] = `
<a
  href="/"
  onClick={[Function]}
>
  <div
    className="css-NavLogo-pmb2010 css-1r3h2fj"
    open={true}
  />
</a>
`;
